codalab/lib/bundle_cli.py

```diff
@@ -1455,43 +1455,50 @@ def copy_bundle(
             },
         )
 
-        # If bundle contents don't exist, finish after just copying metadata
+        # Collect information about how server should unpack
+        filename = nested_dict_get(source_info, 'metadata', 'name')
+        # Fetch bundle contents from source client
         try:
             target_info = source_client.fetch_contents_info(BundleTarget(source_bundle_uuid, ''))
         except NotFoundError:
-            return
+            target_info = None
 
-        # Collect information about how server should unpack
-        filename = nested_dict_get(source_info, 'metadata', 'name')
-        if target_info['type'] == 'directory':
+        # Zip directories
+        if target_info and target_info['type'] == 'directory':
             filename += '.tar.gz'
             unpack = True
         else:
             unpack = False
 
-        # Bundles stuck in non-final states such as 'running' should not keep
-        # that state at the destination server, and should instead just fallback
-        # to 'failed'
-        if source_info['state'] == State.READY:
-            source_state = State.READY
+        try:
+            source_file = source_client.fetch_contents_blob(BundleTarget(source_bundle_uuid, ''))
+        except NotFoundError:
+            source_file = None
+
+        if source_file:
+            # Send file over
+            progress = FileTransferProgress('Copied ', f=self.stderr)
+            with closing(source_file), progress:
+                dest_client.upload_contents_blob(
+                    dest_bundle['id'],
+                    fileobj=source_file,
+                    params={
+                        'filename': filename,
+                        'unpack': unpack,
+                        'simplify': False,  # retain original bundle verbatim
+                        'state_on_success': source_info['state'],  # copy bundle state
+                        'finalize_on_success': True,
+                    },
+                    progress_callback=progress.update,
+                )
         else:
-            source_state = State.FAILED
-
-        # Send file over
-        progress = FileTransferProgress('Copied ', f=self.stderr)
-        source = source_client.fetch_contents_blob(BundleTarget(source_bundle_uuid, ''))
-        with closing(source), progress:
+            # Update the bundle state with final states
             dest_client.upload_contents_blob(
```

Comment:
It looks like we're already calling `upload_contents_blob` above...what is this doing here?

Reply:
I split them into two parts is because `source` is `None` when the bundle doesn't have any content. When bundle has content, `source` is a NOT `None` value, which can be closing.